[Qt] implements concept for different disk sizes on intro

marcoagner opened the pull request:
Fixes https://github.com/bitcoin/bitcoin/issues/13213.
Mostly, I layed out the concept to open the PR for refinement and getting feedback if the approach is okay. Changes are expected.

Two points:
- The values for both new consts `TESTNET_BLOCK_CHAIN_SIZE` and `TESTNET_CHAIN_STATE_SIZE` is certainly not optimal; I just checked the size of my testnet3 related dirs and set them to little bit higher values. Which values should be used?
- Should we do something like this to regtest? Or these "niceties" do not matter when on regtest?

Thanks!

laudaa commented:
Tested on Ubuntu 16.04: [Image](https://i.imgur.com/2DmIUsK.png)
Nano nit: '200GB'/'15GB' should be '200 GB'/'15 GB' for the sake of consistency.

sipa commented:
Thanks for addressing this issue.

We generally try to avoid having explicit branching on the network type. Instead, I would suggest adding a variable to CChainParams (which is initialized differently in the different network versions).

marcoagner commented:
Okay, thanks! I will update this PR with the CChainParams approach.

practicalswift commented:
Concept ACK

jonasschnelli commented:
utACK,... maybe simplify the code according to https://github.com/bitcoin/bitcoin/pull/13229

marcoagner commented:
@jonasschnelli, if #13229 is acceptable, I prefer dropping this PR in favor of #13229. No need to re-do the work here and your approach seems cleaner.

Just as a heads up, it seems that #13229 also explicitly branches the network type (see @sipa's suggestion above). I'm new to the code base and I can't tell for sure if using CChainParams for this Qt issue will be better for the rest of the code as a whole.

MarcoFalke commented:
I think it is fine to put it in chainparams, if you call it `m_assumed_size` (similar to "assumed valid").

jonasschnelli commented:
> @jonasschnelli, if #13229 is acceptable, I prefer dropping this PR in favor of #13229. No need to re-do the work here and your approach seems better.

Please continue with this and feel free to grab code from 13229...

I had the feeling that this would be GUI only and not appropriate to put Into the chainparams holder... but @sipa and @MarcoFalke think its better there which I guess they are right...

sipa commented on doc/release-process.md:
```diff
@@ -23,6 +23,7 @@ Before every major release:
 
 * Update hardcoded [seeds](/contrib/seeds/README.md), see [this pull request](https://github.com/bitcoin/bitcoin/pull/7415) for an example.
 * Update [`BLOCK_CHAIN_SIZE`](/src/qt/intro.cpp) to the current size plus some overhead.
```
This line can be removed now.

marcoagner replied:
ooops, okay

sipa commented on src/chainparams.cpp:
```diff
@@ -317,6 +321,8 @@ class CRegTestParams : public CChainParams {
         pchMessageStart[3] = 0xda;
         nDefaultPort = 18444;
         nPruneAfterHeight = 1000;
+        m_assumed_blockchain_size = 200;
```
This sounds a bit excessive for regtest.

marcoagner replied:
do you have any suggestion for this value? thanks

sipa commented on src/qt/intro.cpp:
```diff
@@ -12,6 +12,7 @@
 
 #include <qt/guiutil.h>
 
+#include <chainparams.h>
```
You can't include core files directly (this is part of ongoing work to separate the two more cleanly); you'll need to go through the node interface (see the include below).

marcoagner replied:
I see! Thank you, I will fix this as soon as I get back.

marcoagner commented:
Updated do reflect @sipa's review.
My only comment for now is that the intro message is a bit funny warning that you need at least 0GB... even though that's accurate. :)
The message didn't reflect regtest reality anyway (e.g. "starting with the earliest transactions in 2009 when Bitcoin initially launched") but let me know if anybody thinks this should be addressed for some reason I can't think of.

marcohextor committed:
implements different disk sizes for different networks on intro

- Creates m_assumed_blockchain_size and m_assumed_chain_state_size on CChainParams.
- Implements access to CChainParams' m_assumed_blockchain_size and m_assumed_chain_state_size on node interface.
- Implements m_assumed_blockchain_size and m_assumed_chain_state_size on qt/intro via node interface.
- Updates release process document with the new CChainParam's values.

marcoagner commented on src/qt/intro.cpp:
```diff
@@ -201,8 +203,15 @@ bool Intro::pickDataDirectory(interfaces::Node& node)
+        /* Use selectParams here to guarantee Params() can be used by node interface */
+        try {
+            node.selectParams(gArgs.GetChainName());
+        } catch(std::exception) {
```
Except this one, other issues were caused by a (obviously bad) merge conflict resolution - bad. I'll take a lot more care so it doesn't happen again.
Thank you for taking the time to review it.

sipa commented:
Concept ACK. Code review ACK for non-Qt code 9d0e52834bbd38e7c7410bcb09ef85d157968b04

jonasschnelli commented:
Re-utACK 9d0e52834bbd38e7c7410bcb09ef85d157968b04

marcoagner commented:
Should anything else be done to have this merged?

jonasschnelli commented:
Tested ACK 9d0e52834bbd38e7c7410bcb09ef85d157968b04

marcoagner commented:
Thank you for testing, @jonasschnelli.

MarcoFalke commented on src/chainparams.cpp:
```diff
@@ -107,6 +107,8 @@ class CMainParams : public CChainParams {
         pchMessageStart[3] = 0xd9;
         nDefaultPort = 8333;
         nPruneAfterHeight = 100000;
+        m_assumed_blockchain_size = 200;
```
This used to be 220?

marcoagner replied:
Well spotted, thanks! This was 200 when the PR was made but changed down the line with this maintenance for 0.17: https://github.com/bitcoin/bitcoin/commit/78dae8caccd82cfbfd76557f1fb7d7557c7b5edb. Do you think this should be addressed here or could wait for the next maintenance (since the release process now asks this to be change)?

MarcoFalke replied:
Could bump it to 240 before 0.18 is branched, so we don't forget about it?

marcoagner replied:
Yes, thanks. Like this? -> https://github.com/bitcoin/bitcoin/pull/15183

MarcoFalke commented on src/qt/intro.cpp:
```diff
@@ -201,8 +199,15 @@ bool Intro::pickDataDirectory(interfaces::Node& node)
+
         /* If current default data directory does not exist, let the user choose one */
-        Intro intro;
+        Intro intro(0, node.getAssumedBlockchainSize(), node.getAssumedChainStateSize());
```
`s/0/nullptr/`

marcoagner replied:
Is another PR for this okay?

maflcko reviewed (commented):
Some post-merge observations
